### PR TITLE
24 create regular expense

### LINKED ISSUE #22: Handle regular expenses
# Card:

**Role:** User
**Goal:** Automatic booking of contracts that generate the same costs in regular periods
**Reason:** The user only has to specify the expense once (easier use of the app)

---

# Confirmation:

## Acceptance tests:

* Regular expenses are added to the user expenses on a specific period (Monthly, Yearly)

### LINKED ISSUE #23: Delete regular expenses
# Card:

**Role:** User
**Goal:** Delete a regular expense.
**Reason:** The user wants to delete expired or non valid regular expenses.

---

# Confirmation:

## Acceptance tests:

* UI (delete regular expense)

### LINKED ISSUE #24: Create regular expense
# Card:

**Role:** User
**Goal:** Specify a monthly or annual expense
**Reason:** The user wants to define an expense that will be made automatically

---

# Confirmation:

## Acceptance tests:

* UI to create the name and amount of the returning expense

### PR DESCRIPTION
**Pull request by elimei1**
You saw most of it already on Discord. Basically you can add a regular transaction with the frequency, on the next "pay" day a normal transaction with the same values gets added.

A few things:
follow-up transactions get only added when you restart the app (but at least all on their correct date)
you have to select a frequency because the calculation for the next date messes up with the default value (don't leave it blank or the nextDueDate is wrong)
A success message for adding a transaction gets displayed, in theory also an error message but here the app crashes before :(
I added the date fields instead of the normal input fields
To delete normal transactions hold the transaction
In order to get no follow-up transactions anymore (delete a regular expense) delete the original (first) regular expense since from this expense all follow-ups are calculated
Changing the frequency afterwards is not possible (only in firebase) since we have no transaction "settings"

Closes #22 
Closes #23 
Closes #24

### CODE REVIEW

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/SimpleDropdownField.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    onItemSelected: (String) -> Unit
+) {
+    var expanded by remember { mutableStateOf(false) }
+    var selectedItem by remember { mutableStateOf<String?>(null) }
+
+    Column {
+        Box(modifier = Modifier.fillMaxWidth().padding(top = 10.dp)) {
+            BasicText(
+                text = selectedItem ?: placeholder,
+                style = MaterialTheme.typography.bodyLarge.copy(color = Color.Black),
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/SimpleDropdownField.kt:35]:
```suggestion
                style = MaterialTheme.typography.bodyLarge.copy(color = MaterialTheme.colorScheme.inverseSurface),
```

Make it work in dark mode.
Before:
Bildschirmfoto 2024-06-26 um 13 33 26
After:
Bildschirmfoto 2024-06-26 um 13 38 48

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/SimpleDropdownField.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    .padding(20.dp)
+                    .fillMaxWidth()
+            )
+            DropdownMenu(
+                expanded = expanded,
+                onDismissRequest = { expanded = false }
+            ) {
+                items.forEach { label ->
+                    DropdownMenuItem(
+                        { BasicText(text = label) },
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/SimpleDropdownField.kt:47]:
```suggestion
                        { Text(text = label, color = MaterialTheme.colorScheme.inverseSurface) },
```
Bildschirmfoto 2024-06-26 um 13 46 55

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    OutlinedTextField(
+        value = "${selectedDate.dayOfMonth} ${selectedDate.month} ${selectedDate.year}",
+        onValueChange = {},
+        modifier = Modifier.clickable {
+            showDialog.value = true
+        },
+        enabled = false,
+        label = { Text("Date", color = Color.Black) },
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:71]:
```suggestion
        label = { Text("Date") },
```

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        textStyle = TextStyle(color = Color.Black, fontSize = 16.sp),
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:72]:
```suggestion
        textStyle = TextStyle(color = MaterialTheme.colorScheme.onSurfaceVariant, fontSize = 16.sp),
```

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        colors = TextFieldDefaults.outlinedTextFieldColors(
+            focusedBorderColor = Color.Black,
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:74]:
```suggestion
            focusedBorderColor = MaterialTheme.colorScheme.onSurfaceVariant,
```

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            unfocusedBorderColor = Color.Black,
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:75]:
```suggestion
            unfocusedBorderColor = MaterialTheme.colorScheme.onSurfaceVariant,
```

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            disabledBorderColor = Color.Black
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:76]:
```suggestion
            disabledBorderColor = MaterialTheme.colorScheme.onSurfaceVariant,
```

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        )
+    )
+}
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionDateField.kt:79]: Bildschirmfoto 2024-06-26 um 13 54 54
Bildschirmfoto 2024-06-26 um 13 55 24

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionListItem.kt
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionListItem.kt:1]: Idea: Add badge that says "first transaction of regular ..."

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/screens/TransactionScreen.kt [hunk 9]
```diff
@@ -9,22 +9,39 @@ import androidx.compose.material.icons.automirrored.filled.ArrowBack
...
     navigationController: NavController,
     viewModel: UserViewModel
 ) {
+
+    val transactionMessage by viewModel.transactionMessage
+    var transactionDate by rememberSaveable { mutableStateOf(LocalDateTime.now()) }
+
     Scaffold(
         topBar = {
             SimpleTopAppBar(title = "Add transaction") {
```
Comment (michael-mezgo) [inline app/src/main/java/at/ac/fhcampuswien/budget_fox/screens/TransactionScreen.kt:47]: Pls use other button that works in dark mode
```Kotlin
        topBar = {
            SimpleTopAppBar(
                title = "Join a Household"
            ) {
                IconButton(onClick = {
                    navigationController.popBackStack()
                }) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                        contentDescription = "Go back"
                    )
                }
            }
        }
```

**Review by michael-mezgo — CHANGES_REQUESTED** [review]

**Review by EP1210 — APPROVED** [review]
Adding a regular expense seems to work flawlessly. Thank you for that :+1:. A possible improvement for the future would be that follow up transactions are displayed immediately if that is somehow possible. Also some overall exception handling when adding a transaction might be helpful to prevent the app from crashing.

Apart from what Michael already mentioned (text field or navigation icon is not visible in dark mode) I have no complaints.

**Comment by EP1210** [comment]
Now everything is nicely visible in dark mode :+1:.

File: app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionListItem.kt
Comment (elimei1) [reply app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionListItem.kt:1]: Like that?
Or what do you mean?
Screenshot_20240626-221902

Comment (michael-mezgo) [reply app/src/main/java/at/ac/fhcampuswien/budget_fox/widgets/TransactionListItem.kt:1]: Yeah. Perfect 👍

**Review by michael-mezgo — APPROVED** [review]
Gerat job 🥇.
The concept for the badge is also top 🔝 (pls add it).

**Comment by michael-mezgo** [comment]
@elimei1 add the badge to #39

**Comment by michael-mezgo** [comment]
> @elimei1 add the badge to #39

Solved. Thanks @elimei1 👍

**Pull request merged**